Unhandled `llvm.assume` intrinsic

Given the following C code:
```c
#include <stdint.h>
void test13(uint32_t var15, int8_t var2[8]) {
  for (uint32_t var4 = var15 | 0 >= 0; var4 < (0 <= 0); var4 += 1)
    var2[0] = 0;
}
#ifdef HLS_FUZZER_VERIFY
constexpr
#endif
    void
    test_bench() {
}
int main() {}
```
The import from LLVM IR fails due to a missing intrinsic function handler:
```
dynamatic> compile
[INFO] Copied C source
/home/mboeck/dynamatic/build/bin/bug20/out/comp/test13.c:3:30: warning: | has lower precedence than >=; >= will be evaluated first [-Wparentheses]
    3 |   for (uint32_t var4 = var15 | 0 >= 0; var4 < (0 <= 0); var4 += 1)
      |                              ^~~~~~~~
/home/mboeck/dynamatic/build/bin/bug20/out/comp/test13.c:3:30: note: place parentheses around the '>=' expression to silence this warning
    3 |   for (uint32_t var4 = var15 | 0 >= 0; var4 < (0 <= 0); var4 += 1)
      |                              ^       
      |                                (     )
/home/mboeck/dynamatic/build/bin/bug20/out/comp/test13.c:3:30: note: place parentheses around the | expression to evaluate it first
    3 |   for (uint32_t var4 = var15 | 0 >= 0; var4 < (0 <= 0); var4 += 1)
      |                              ^
      |                        (        )
1 warning generated.
[INFO] Disabled short-circuiting
/home/mboeck/dynamatic/build/bin/bug20/out/comp/test13.c:3:30: warning: | has lower precedence than >=; >= will be evaluated first [-Wparentheses]
    3 |   for (uint32_t var4 = var15 | 0 >= 0; var4 < (0 <= 0); var4 += 1)
      |                              ^~~~~~~~
/home/mboeck/dynamatic/build/bin/bug20/out/comp/test13.c:3:30: note: place parentheses around the '>=' expression to silence this warning
    3 |   for (uint32_t var4 = var15 | 0 >= 0; var4 < (0 <= 0); var4 += 1)
      |                              ^       
      |                                (     )
/home/mboeck/dynamatic/build/bin/bug20/out/comp/test13.c:3:30: note: place parentheses around the | expression to evaluate it first
    3 |   for (uint32_t var4 = var15 | 0 >= 0; var4 < (0 <= 0); var4 += 1)
      |                              ^
      |                        (        )
1 warning generated.
[INFO] Compiled to LLVM IR
[INFO] Optimized LLVM IR
[INFO] Applied memory dependency analysis to LLVM IR
LLVM ERROR: Not implemented llvm intrinsic function handling!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /home/mboeck/dynamatic/build/bin/translate-llvm-to-std /home/mboeck/dynamatic/build/bin/bug20/out/comp/clang.opt.dep.ll -function-name test13 -csource /home/mboeck/dynamatic/build/bin/bug20/test13.c -dynamatic-path /home/mboeck/dynamatic -o /home/mboeck/dynamatic/build/bin/bug20/out/comp/cf.mlir
 #0 0x00000000006192d8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/mboeck/dynamatic/build/bin/translate-llvm-to-std+0x6192d8)
 #1 0x0000000000616f5e llvm::sys::RunSignalHandlers() (/home/mboeck/dynamatic/build/bin/translate-llvm-to-std+0x616f5e)
 #2 0x0000000000619bed SignalHandler(int) (/home/mboeck/dynamatic/build/bin/translate-llvm-to-std+0x619bed)
 #3 0x00007fe14b37a990 __restore_rt (/lib64/libpthread.so.0+0x12990)
 #4 0x00007fe14aa335af raise (/lib64/libc.so.6+0x4e5af)
 #5 0x00007fe14aa06ee5 abort (/lib64/libc.so.6+0x21ee5)
 #6 0x00000000005f2a88 llvm::report_fatal_error(llvm::Twine const&, bool) (/home/mboeck/dynamatic/build/bin/translate-llvm-to-std+0x5f2a88)
 #7 0x00000000005f2896 (/home/mboeck/dynamatic/build/bin/translate-llvm-to-std+0x5f2896)
 #8 0x0000000000599770 void TranslateLLVMToStd::naiveTranslation<mlir::arith::NegFOp>(mlir::Type, mlir::ValueRange, llvm::Instruction*) (/home/mboeck/dynamatic/build/bin/translate-llvm-to-std+0x599770)
 #9 0x00000000005932b7 TranslateLLVMToStd::translateFunction(llvm::Function*) (/home/mboeck/dynamatic/build/bin/translate-llvm-to-std+0x5932b7)
#10 0x0000000000592239 TranslateLLVMToStd::translateLLVMModule() (/home/mboeck/dynamatic/build/bin/translate-llvm-to-std+0x592239)
#11 0x0000000000590b03 main (/home/mboeck/dynamatic/build/bin/translate-llvm-to-std+0x590b03)
#12 0x00007fe14aa1f865 __libc_start_main (/lib64/libc.so.6+0x3a865)
#13 0x000000000058fc3e _start (/home/mboeck/dynamatic/build/bin/translate-llvm-to-std+0x58fc3e)
/home/mboeck/dynamatic/tools/dynamatic/scripts/compile.sh: line 217: 1604259 Aborted                 (core dumped) $LLVM_TO_STD_TRANSLATION_BIN "$F_CLANG_OPTIMIZED_DEPENDENCY" -function-name "$KERNEL_NAME" -csource "$F_C_SOURCE" -dynamatic-path "$DYNAMATIC_DIR" -o "$F_CF"
[FATAL] Failed to convert to std dialect
```
Looking at the generated LLVM IR the issue is seemingly the presence of `llvm.assume`:
```llvm
; Function Attrs: nounwind uwtable
define dso_local void @test13(i32 noundef %var15, ptr noundef %var2) #0 {
entry:
  %or = or i32 %var15, 1
  %cmp = icmp eq i32 %or, 0
  %0 = xor i1 %cmp, true
  call void @llvm.assume(i1 %0)
  ret void
}
```
The intrinsic ought to be safe to drop since it only encodes extra optimization information

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -11,6 +11,7 @@ set(DYNAMATIC_TEST_DEPENDS
   dynamatic-opt
   export-rtl
   hls-fuzzer-check-bitwidth
+  translate-llvm-to-std
   source-rewriter
   DynPragmasPlugin
   )

--- a/test/lit.cfg.py
+++ b/test/lit.cfg.py
@@ -16,7 +16,7 @@ config.name = "Dynamatic"
 config.test_format = lit.formats.ShTest(not llvm_config.use_lit_shell)
 
 # suffixes: A list of file extensions to treat as test files.
-config.suffixes = [".mlir"]
+config.suffixes = [".mlir", ".ll"]
 
 # test_source_root: The root path where tests are located.
 config.test_source_root = os.path.dirname(__file__)
@@ -48,6 +48,8 @@ tools = ["dynamatic-opt", "hls-fuzzer-check-bitwidth",
                    command=f"cp %s %t.c && {config.dynamatic_tools_dir}/source-rewriter %t.c --"),
          ToolSubst("%export-vhdl",
                    command=f"rm -rf %t; mkdir %t; {config.dynamatic_tools_dir}/export-rtl %s %t {config.dynamatic_src_root}/data/rtl-config-vhdl.json --dynamatic-path {config.dynamatic_src_root} --hdl vhdl"),
+         ToolSubst("%translate-llvm-to-std",
+                   command=f"{config.llvm_tools_dir}/split-file %s %t; {config.dynamatic_tools_dir}/translate-llvm-to-std %t/test.ll -csource %t/test.c -function-name=test --dynamatic-path {config.dynamatic_src_root}"),
          ToolSubst("%dyn-clang-pragmas",
                    command=f"{config.llvm_tools_dir}/clang "
                    f"-fplugin={config.dynamatic_shlib_dir}/DynPragmasPlugin{config.llvm_shlib_ext} "

--- a/test/tools/translate-llvm-to-std/llvm-assume.ll
+++ b/test/tools/translate-llvm-to-std/llvm-assume.ll
@@ -1,0 +1,16 @@
+; RUN: %translate-llvm-to-std -o - | FileCheck %s
+
+;--- test.ll
+
+; CHECK: func.func @test(
+define void @test(i16 %var0) {
+  %t = trunc i16 %var0 to i1
+  call void @llvm.assume(i1 %t)
+  ret void
+}
+
+declare void @llvm.assume(i1 noundef)
+
+;--- test.c
+
+void test(short var0);

--- a/tools/translate-llvm-to-std/TranslateLLVMToStd.cpp
+++ b/tools/translate-llvm-to-std/TranslateLLVMToStd.cpp
@@ -1021,6 +1021,9 @@ void TranslateLLVMToStd::translateCallInst(llvm::CallInst *callInst) {
   } else if (calledFunc->getIntrinsicID() == Intrinsic::fshl ||
              calledFunc->getIntrinsicID() == Intrinsic::fshr) {
     this->translateFunnelShiftIntrinsic(callInst);
+  } else if (calledFunc->getIntrinsicID() == Intrinsic::assume) {
+    // An assume op can be simply dropped. It does not have any semantics
+    // besides causing undefined behavior.
   } else {
     llvm::report_fatal_error(
         "Not implemented llvm intrinsic function handling!");

